File: Striver Important Interview Questions/Day 9 - Recursion/subsets.cpp

// https://leetcode.com/problems/subsets/

class Solution {
public:

    void solve(vector<int>&nums, vector<vector<int>>&ans, vector<int>&subset, int startingIndex){
        ans.push_back(subset);

        for(int i = startingIndex; i < nums.size(); i++){
            subset.push_back(nums[i]);
            solve(nums, ans, subset, i+1);
            subset.pop_back();
        }

        return;
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        int n = nums.size();
        if(n == 0) return ans;

        vector<int> subset;
        int startingIndex = 0;
        solve(nums, ans, subset, startingIndex);

        return ans;
    }
};

File: Neetcode-150-js/Backtracking/subsets.js

// https://leetcode.com/problems/subsets/description/

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    let ans = [];
    let len = nums.length;
    if(len === 0) return ans;
    
    const subset = [];

    var solve = (si) => {
        ans = [...ans, [...subset]];

        for(let i = si; i < len; i++){
            subset.push(nums[i]);
            solve(i+1);
            subset.pop();
        }
    }

    let si = 0;
    solve(si);
    return ans;
};